lib/lpc55-puf/src/lib.rs

```diff
@@ -0,0 +1,237 @@
+    }
+
+    pub fn new(puf: &'a PUF) -> Self {
+        Self { puf }
+    }
+
+    /// Generate a new key code for a key with the provided PUF index &
+    /// length.
+    /// NOTE: The PUF doesn't return the key immediately. Instead it
+    /// returns a keycode through the second param. This keycode can later
+    /// be used to create and return the key to the caller.
+    pub fn generate_keycode(
+        &self,
+        index: u32,
+        key_len: usize,
+        keycode: &mut [u32],
+    ) -> bool {
+        if !self.is_generatekey_allowed() {
+            panic!("PufCmdDisallowed");
+        }
+
+        // devide by sizeof u32 here because keycode param is an array of u32
+        let keycode_len =
+            Self::key_to_keycode_len(key_len) / mem::size_of::<u32>();
+        if keycode.len() < keycode_len {
+            panic!("PufKeyCode");
+        }
+
+        self.set_key_index(index);
+        self.set_key_size(key_len);
+
+        self.puf.ctrl.write(|w| w.generatekey().set_bit());
+        if !self.wait_for_cmd_accept() {
+            panic!("PufCmdAccept");
+        }
+
+        // while PUF is busy, read out whatever part of the KC is available
+        let mut idx = 0;
+        while self.is_busy() {
+            if idx > keycode.len() - 1 {
+                panic!("PufKCTooLong");
+            }
+            if self.is_keycode_part_avail() {
+                let keycode_part = self.puf.codeoutput.read().bits();
+                keycode[idx] = keycode_part;
+                idx += 1;
+            }
+        }
+
+        self.is_success()
+    }
+
+    pub fn get_key(&self, keycode: &[u32], key: &mut [u8]) -> bool {
+        if !self.is_getkey_allowed() {
+            panic!("DenyGetKey");
+        }
+
+        // execute CTRL function / set GETKEY bit in CTRL register, no params
+        self.puf.ctrl.write(|w| w.getkey().set_bit());
+
+        self.wait_for_cmd_accept();
+
+        let mut kc_idx = 0;
+        let mut key_idx = 0;
+        // while PUF busy ... error detection?
+        while self.is_busy() && !self.is_error() {
+            if self.is_keycode_part_req() {
+                self.puf
+                    .codeinput
+                    .write(|w| unsafe { w.bits(keycode[kc_idx]) });
+                kc_idx += 1;
+            }
+            if self.is_key_part_avail() {
+                for byte in self.puf.keyoutput.read().bits().to_ne_bytes() {
+                    key[key_idx] = byte;
+                    key_idx += 1;
+                }
+            }
+        }
+
+        self.is_success()
+    }
+
+    // Set key index (between 0 & 15) for a key generated by the PUF or set
+    // through the API. This value is ignored for the GetKey command as the index
+    // is baked into the KeyCode.
+    // NOTE: The only use for the key index is to send keys to hardware blocks
+    // like the AES & PRINCE hardware (key index 0). Any key index != 0 is fine
+    // for our purposes.
```

Comment:
Seems reasonable to me. I'm working up the requested comment now.

I haven't looked at the lock registers much but just to make sure I understand the situation correctly: The NXP ROM uses key index 15 for the UDS key code and prevent usage of keycode 15 by setting `IDX15` to `0x1`. The ROM locks this whole register by setting `LOCK_IDX` to `0x1`. This causes the situation you describe above in that this prevents the use of index 15 but also prevents us from disabling the use of the other indexes controlled by this register. So NXP has effectively locked key index 8 - 14 as active / usable.

It looks like the `IDXBLK_L_DP` and `IDXBLK_H_DP` registers may have been intended to mitigate this though w/o the locking mechanism which AFAIK makes them less than useful? We could disable indexes 8 - 14 by changing `IDXBLK_H_DP` so that it disagrees with `IDXBLK_H` which according to the docs will prevent use of the index, but only until someone changes `IDXBLK_H_DP` back since it cannot be locked like `IDXBLK_H`. I'm not sure there's any point in doing this since it's effectively "opt in / out" but maybe I'm missing something?

Reply:
comment added

Reply:
The `DP` registers are suppose to be there to avoid glitch attacks. I'm not sure what happens if you try to modify one of them when it's already locked but it's a worthwhile experiment.

Reply:
I've been poking at these and both DP registers are `0u32`. According to the NXP docs if the index blocking register and it's DP counter part disagree about a given index then the index should be disabled. Sooooo (unless I'm screwing up my experiment) AFAIK the DP registers don't do anything :(

Reply:
update: though not R/W (like the manual says) the DP registers do indeed affect access to a given index if the relevant bits in the DP and its counterpart disagree. The puf module in this PR has been changed such that updates made to the IDXBLK registers are also written to the associated DP register:
* https://github.com/oxidecomputer/hubris/pull/1063/files#diff-17976721289814e9516272c40da1540fdbc0ad32ea1191a1f0221e0f2239fca8R230
* https://github.com/oxidecomputer/hubris/pull/1063/files#diff-17976721289814e9516272c40da1540fdbc0ad32ea1191a1f0221e0f2239fca8R239